### PR TITLE
Updates sklearn to version 0.21; closes #629

### PR DESCRIPTION
**Pull request by gui-miotto**
* Eliminates the dependency on xgb
* Metalearning tables need to be updated

### CODE REVIEW

File: autosklearn/pipeline/components/classification/gradient_boosting.py [hunk 1]
```diff
@@ -125,34 +93,25 @@ def get_properties(dataset_properties=None):
     @staticmethod
     def get_hyperparameter_search_space(dataset_properties=None):
         cs = ConfigurationSpace()
-        loss = Constant("loss", "deviance")
+        loss = Constant("loss", "auto")
```
Comment (mfeurer) [inline autosklearn/pipeline/components/classification/gradient_boosting.py]: Could you please order the hyperparameters consistent with the constructor? This will make it easier to follow the code.

File: autosklearn/pipeline/components/classification/gradient_boosting.py [hunk 2]
```diff
@@ -125,34 +93,25 @@ def get_properties(dataset_properties=None):
...
-            name='min_impurity_decrease', value=0.0)
-        cs.add_hyperparameters([loss, learning_rate, n_estimators, max_depth,
-                                criterion, min_samples_split, min_samples_leaf,
-                                min_weight_fraction_leaf, subsample,
-                                max_features, max_leaf_nodes,
-                                min_impurity_decrease])
+        max_iter = UniformIntegerHyperparameter(
+            "max_iter", 50, 500, default_value=100)
+        max_bins = UniformIntegerHyperparameter(
+            name="max_bins", lower=2, upper=256, default_value=256, log=False)
```
Comment (mfeurer) [inline autosklearn/pipeline/components/classification/gradient_boosting.py]: I think we shouldn't tune this value as it mostly covers the training speed which is not an optimization objective of Auto-sklearn.

File: autosklearn/pipeline/components/regression/gradient_boosting.py [hunk 3]
```diff
@@ -123,34 +94,24 @@ def get_properties(dataset_properties=None):
...
-        cs.add_hyperparameters([loss, learning_rate, n_estimators, max_depth,
-                                min_samples_split, min_samples_leaf,
-                                min_weight_fraction_leaf, subsample, max_features,
-                                max_leaf_nodes, min_impurity_decrease, alpha])
+        cs.add_hyperparameters([loss, learning_rate, max_iter, max_depth, 
+                                min_samples_leaf, max_leaf_nodes, max_bins, 
+                                l2_regularization])
 
-        cs.add_condition(InCondition(alpha, loss, ['huber', 'quantile']))
         return cs
```
Comment (mfeurer) [inline autosklearn/pipeline/components/regression/gradient_boosting.py]: The code for both appears to be almost identical. I think we should make it use a common base class to avoid code duplication.

File: autosklearn/__init__.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 __MANDATORY_PACKAGES__ = '''
 numpy>=1.9
-scikit-learn>=0.19,<0.20
+scikit-learn==0.21.0
```
[the remark this answers is not in the document]
Comment (gui-miotto) [reply autosklearn/__init__.py]: ok

File: .gitignore [hunk 5]
```diff
@@ -49,3 +52,4 @@ num_run
 number_submission
 .pypirc
 dmypy.json
+*.log
```
[the remark this answers is not in the document]
Comment (gui-miotto) [reply .gitignore]: So, ignore all .c files except `competition_c_functions.c`?

[the remark this answers is not in the document]
Comment (mfeurer) [reply .gitignore]: No, please ignore all of them. Also `competition_c_functions.c` as this is generated automatically.

File: setup.py [hunk 6]
```diff
@@ -43,6 +43,27 @@ def finalize_options(self):
 ]
 
 
+requirements = [
```
[the remark this answers is not in the document]
Comment (gui-miotto) [reply setup.py]: Ok. (This one was already there in the branch that I took over)

File: autosklearn/pipeline/components/classification/gradient_boosting.py (hunk 1 above)
Comment (gui-miotto) [reply autosklearn/pipeline/components/classification/gradient_boosting.py]: ok

File: autosklearn/pipeline/components/regression/gradient_boosting.py (hunk 3 above)
Comment (gui-miotto) [reply autosklearn/pipeline/components/regression/gradient_boosting.py]: Just for the record, we decided to don't do this one for now.

File: autosklearn/pipeline/components/classification/gradient_boosting.py (hunk 2 above)
Comment (gui-miotto) [reply autosklearn/pipeline/components/classification/gradient_boosting.py]: ok

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (gui-miotto) [reply .gitignore]: Oh, no, wait. I think you want me to untrack competition_c_functions.c, right? 
(*.c is already in .gitignore)

[the remark this answers is not in the document]
Comment (mfeurer) [reply .gitignore]: If *.c is already in gitignore, then yes, I'd like you to untrack that file.

File: test/test_pipeline/components/feature_preprocessing/test_kernel_pca.py [hunk 7]
```diff
@@ -36,11 +37,11 @@ def test_default_configuration_classify(self):
             X_test_trans = preprocessor.transform(X_test)
```
Comment (gui-miotto) [inline test/test_pipeline/components/feature_preprocessing/test_kernel_pca.py]: @mfeurer : pushed this fix for the kernel PCA test. Although there were two missing random_state assignments, I think the main problem was caused by this line. It was having numerical issues and returning non-deterministic `X_test_trans`. This problem does not occur with other datasets (e.g. iris) or if we use 'digits' with a larger training dataset. So I increased it to 1000 samples.

File: test/test_pipeline/test_classification.py [hunk 8]
```diff
@@ -206,7 +206,7 @@ def test_configurations_categorical_data(self):
             dataset_properties={'sparse': False},
             include={
                 'preprocessor': ['no_preprocessing'],
-                'classifier': ['sgd', 'gradient_boosting']
+                'classifier': ['sgd', 'adaboost']
```
Comment (gui-miotto) [inline test/test_pipeline/test_classification.py]: @mfeurer : this test was failing because the new gradient_boosting currently doesn't support sample_weight. Since this is a test of the classification pipeline (and not of a specific classifier), I thought that the quickest fix would be to just pick another classifier with support for sample_weight. Hope that's alright.

Comment (mfeurer) [reply test/test_pipeline/test_classification.py]: Yes, that makes sense.

File: test/test_evaluation/test_train_evaluator.py [hunk 9]
```diff
@@ -612,8 +613,11 @@ def test_subsample_indices_regression(self, mock, backend_mock):
 
         # Corner cases
         evaluator.subsample = 0
-        train_indices5 = evaluator.subsample_indices(train_indices)
-        np.testing.assert_allclose(train_indices5, np.array([]))
```
Comment (gui-miotto) [inline test/test_evaluation/test_train_evaluator.py]: This behavior has changed in sklearn 0.21. Now a train_size=0 will raise an exception. Does autosklearn make any assumptions for this case? In other words, do we have to edit TrainEvaluator.subsample_indices() to return np.array([]) when self.subsample=0?

File: test/test_pipeline/components/feature_preprocessing/test_kernel_pca.py (hunk 7 above)
Comment (mfeurer) [reply test/test_pipeline/components/feature_preprocessing/test_kernel_pca.py]: This appears to be an open issue: https://github.com/scikit-learn/scikit-learn/issues/5970

**Comment by gui-miotto** [comment]
@mfeurer : All tests green on my computer. Time to generate the new metalearning?

**Comment by gui-miotto** [comment]
@mfeurer , everything (except metalearning) passing on Travis-CI

File: test/test_pipeline/components/classification/test_passive_aggressive.py [hunk 10]
```diff
@@ -11,16 +11,16 @@ class PassiveAggressiveComponentTest(BaseClassificationComponentTest):
...
 
     res = dict()
-    res["default_iris"] = 0.78
-    res["iris_n_calls"] = 4
-    res["default_iris_iterative"] = 0.98
-    res["default_iris_proba"] = 0.4471909030527266
-    res["default_iris_sparse"] = 0.36
-    res["default_digits"] = 0.9253187613843351
-    res["digits_n_calls"] = 4
+    res["default_iris"] = 0.92
```
Comment (amueller) [inline test/test_pipeline/components/classification/test_passive_aggressive.py]: Do you know what caused these changed?

**Comment by amueller** [comment]
awesome!!

File: test/test_pipeline/components/classification/test_passive_aggressive.py (hunk 10 above)
Comment (mfeurer) [reply test/test_pipeline/components/classification/test_passive_aggressive.py]: Unfortunately not. Maybe the iterative training was one wrong before?
